### PR TITLE
GROW-162 - Relay Metrics Aggregation

### PR DESCRIPTION
**Pull request by tabakuj**
* **Please check if the PR fulfills these requirements**
- [ ] Commit message follows the Contribution Guidelines
- [x] Tests ran locally and added/modified if needed
- [ ] Docs have been added/updated, if applicable
- [ ] If applicable - JIRA ticket ID was added

* **What kind of change does this PR introduce?** (Bug fix, feature, unit tests, docs update, ...)

Adding relay metrics
* **What is the current behavior?** (You can also link to an open issue here)
nothing is saved

* **What is the new behavior (if this is a feature change)?**

* **Please describe what manual tests you ran, if applicable**

* **Other information**:

### CODE REVIEW

File: relayer/chainproxy/chainproxy.go [hunk 1]
```diff
@@ -26,6 +27,7 @@ const (
 	ContextUserValueKeyDappID = "dappID"
 	MinimumTimePerRelayDelay  = time.Second
 	AverageWorldLatency       = 200 * time.Millisecond
+	ContextUserValueAnalytics = "analytics"
```
Comment (nirtayeb) [inline relayer/chainproxy/chainproxy.go:30]: please clean it (probably it left from my previous code)

File: relayer/chainproxy/grpc.go [hunk 2]
```diff
@@ -279,7 +279,8 @@ func (cp *GrpcChainProxy) PortalStart(ctx context.Context, privKey *btcec.Privat
 		msgSeed := cp.portalLogs.GetMessageSeed()
 		utils.LavaFormatInfo("GRPC Got Relay: "+method, nil)
 		var relayReply *pairingtypes.RelayReply
-		if relayReply, _, err = SendRelay(ctx, cp, privKey, method, string(reqBody), "", "NoDappID"); err != nil {
+
+		if relayReply, _, err = SendRelay(ctx, cp, privKey, method, string(reqBody), "", "NoDappID", nil); err != nil {
```
Comment (nirtayeb) [inline relayer/chainproxy/grpc.go:283]: we should have analytics on grpc also

File: relayer/chainproxy/portalLogs.go [hunk 3]
```diff
@@ -85,3 +91,10 @@ func (cp *PortalLogs) LogStartTransaction(name string) {
 		defer txn.End()
 	}
 }
+
+func (cp *PortalLogs) AddMetric(data *metrics.RelayAnalytics, IsNotSuccessful bool) {
```
Comment (nirtayeb) [inline relayer/chainproxy/portalLogs.go:95]: cp is confusing name for PortalLogs

File: relayer/metrics/analytics.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+type RelayAnalytics struct {
+	ProjectHash string // imported - REST
+	Timestamp   time.Time
+	ChainID     string /// DONE - SendRelay
+	APIType     string /// DONE - SendRelay
+	// latency in ms
+	Latency      int64  // DONE - SendRelay
+	Success      bool   // Done - SendRelay
+	ComputeUnits uint64 // Done - SendRelay
```
Comment (nirtayeb) [inline relayer/metrics/analytics.go:15]: Remove my comments here and below too

File: relayer/metrics/analytics.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package metrics
+
+import (
+	"time"
+)
+
+type RelayAnalytics struct {
```
Comment (nirtayeb) [inline relayer/metrics/analytics.go:7]: RelayMetrics may be a better name

File: relayer/metrics/metricsService.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+type MetricService struct {
+	numberOfRelays      int64
+	aggregatedMetricMap *map[string]map[string]map[string]*AggregatedMetric
+	metricsChannel      chan RelayAnalytics
+	sqsUrl              string
+}
+
+func NewMetricService() *MetricService {
+	metricPortalSqsUrl := os.Getenv("METRICS_PORTAL_SQS_URL")
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:28]: REPORT_METRICS_URL

File: relayer/metrics/metricsService.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	intervalData := os.Getenv("METRICS_INTERVAL_FOR_SENDING_DATA_INM")
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:29]: What is INM ?

File: relayer/metrics/metricsService.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+		aggregatedMetricMap: &map[string]map[string]map[string]*AggregatedMetric{},
+	}
+	//setup reader
+	go result.ReadFromChannel()
+	//setup sending of the results into the query
+	ticker := time.NewTicker(time.Duration(intervalForMetrics * time.Minute.Nanoseconds()))
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:50]: we should have here another case for reading from the channel.
so when it ticks we don't read, and when it reads from the channel we don't send data

Comment (nirtayeb) [reply relayer/metrics/metricsService.go:50]: see last example here 
https://dev.to/julianchu/go-for-select-with-timer-1ah8

Comment (nirtayeb) [reply relayer/metrics/metricsService.go:50]: right now you have a race condition between adding metrics and reseting the map

File: relayer/chainproxy/portalLogs.go [hunk 9]
```diff
@@ -19,28 +20,33 @@ const webSocketCloseMessage = "websocket: close 1005 (no status)"
 
 type PortalLogs struct {
 	newRelicApplication *newrelic.Application
+	MetricService       *metrics.MetricService
+	StoreMetricData     bool
 }
 
-func NewPortalLogs() (*PortalLogs, error) {
+func NewPortalLogs(storeMetricData bool) (*PortalLogs, error) {
```
Comment (nirtayeb) [inline relayer/chainproxy/portalLogs.go:27]: it should be decided based on environment variable.

File: relayer/metrics/metricsService.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+	TotalLatency int64
+	RelaysCount  uint64
+	SuccessCount int64
+}
+
+type MetricService struct {
+	numberOfRelays      int64
+	aggregatedMetricMap *map[string]map[string]map[string]*AggregatedMetric
+	metricsChannel      chan RelayAnalytics
+	sqsUrl              string
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:24]: reportUrl

File: relayer/metrics/metricsService.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+		return fmt.Errorf("couldn't send the data into the server")
+	}
+	return nil
+}
+
+func (m *MetricService) ConvertDataToArray() []RelayAnalyticsDTO {
+	if m.numberOfRelays == 0 || m.aggregatedMetricMap == nil {
+		return nil
+	}
+	toSendData := make([]RelayAnalyticsDTO, m.numberOfRelays)
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:86]: if you use append() you don't have to accumulate the numberOfRelays, and have one less dependency/variable to clear

File: relayer/metrics/metricsService.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+		}
+	}()
+	return result
+}
+
+func (m *MetricService) SendTo(data []RelayAnalyticsDTO) error {
+	if data == nil || len(data) == 0 {
+		return nil
+	}
+	jsonValue, err := json.Marshal(data)
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:68]: maybe it is premature optimization, but I think we should send each project data by itself. 
with big messages there is higher chance to lost data on the wire / make queue consumption later slower.

File: relayer/metrics/metricsService.go [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+			if data.Success {
+				toInsertData.SuccessCount = 1
+			}
+			m.StoreAggregatedData(data.ProjectHash, data.ChainID, data.APIType, toInsertData)
+
+		}
+	}
+}
+
+func (m *MetricService) StoreAggregatedData(projectHash string, chainId string, apiType string, data *AggregatedMetric) error {
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:142]: should be private function ("storeAggregatedData")

File: relayer/metrics/metricsService.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+	}
+}
+
+func (m *MetricService) SendDataToChannel(data RelayAnalytics) {
+	if m.metricsChannel != nil {
+		m.metricsChannel <- data
+	}
+}
+
+func (m *MetricService) ReadFromChannel() {
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:121]: should be private

File: relayer/metrics/metricsService.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+	return toSendData
+}
+
+func (m *MetricService) CloseChannel() {
+	if m.metricsChannel != nil {
+		m.CloseChannel()
+	}
+}
+
+func (m *MetricService) SendDataToChannel(data RelayAnalytics) {
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:115]: better naming should be just send / add, the caller of the function don't need to know about internal implementation

File: relayer/metrics/metricsService.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+				counter++
+			}
+		}
+	}
+	m.aggregatedMetricMap = &map[string]map[string]map[string]*AggregatedMetric{}
+	m.numberOfRelays = 0
+	return toSendData
+}
+
+func (m *MetricService) CloseChannel() {
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:109]: seems that nobody calls it

File: relayer/metrics/metricsService.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+				continue
+			}
+			toInsertData := &AggregatedMetric{
+				TotalLatency: data.Latency,
+				RelaysCount:  data.ComputeUnits,
+				SuccessCount: 0,
+			}
+			if data.Success {
+				toInsertData.SuccessCount = 1
+			}
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:135]: why are you convert to this object if it's the same? hust use the RelayAnalytics as the channel source.

I started to wrote the AggregatedMetric for the final aggregation (Totals of many metrics together)

**Comment by nirtayeb** [comment]
Please add unit tests and fix the lint also

File: relayer/metrics/metricsService.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+	counter := 0
+	//project-chain-apiType
+	for projectKey, projectData := range *m.aggregatedMetricMap {
+		for chainKey, chainData := range projectData {
+			for apiTypekey, apiTypeData := range chainData {
+				toSendData[counter] = RelayAnalyticsDTO{
+					ProjectHash:  projectKey,
+					APIType:      apiTypekey,
+					ChainID:      chainKey,
+					Latency:      apiTypeData.TotalLatency,
```
Comment (nirtayeb) [inline relayer/metrics/metricsService.go:96]: here we need to divide the total latency by the relays count

File: relayer/chainproxy/portalLogs.go (hunk 9 above)
Comment (tabakuj) [reply relayer/chainproxy/portalLogs.go:27]: that is for differentiating btw server and portal_server, I added support for enabling or disabling this

File: relayer/metrics/metricsService.go (hunk 7 above)
Comment (tabakuj) [reply relayer/metrics/metricsService.go:29]: in minutes :P

File: relayer/metrics/metricsService.go (hunk 12 above)
Comment (tabakuj) [reply relayer/metrics/metricsService.go:68]: it should be ok for us, since we group the data.

File: relayer/metrics/metricsService.go (hunk 17 above)
Comment (tabakuj) [reply relayer/metrics/metricsService.go:135]: this is not the same. here we store the count and success count which is different from what we are getting

File: relayer/chainproxy/portalLogs.go (hunk 9 above)
Comment (nirtayeb) [reply relayer/chainproxy/portalLogs.go:27]: we may want to support it also for providers ("server") in the future, and until then we can just use the new env var to disable it. please remove the variable from the constructor, it's redundant with the env var

File: relayer/metrics/metricsService.go (hunk 12 above)
Comment (nirtayeb) [reply relayer/metrics/metricsService.go:65]: I prefer it separated, so if we have an issue during processing of one projects metrics we won't loose all the rest.

**Comment by nirtayeb** [comment]
LGTM, Unit testing are on the way.
@ranlavanet please take a look

**Review by nirtayeb — APPROVED** [review]

File: relayer/chainproxy/rest.go [hunk 19]
```diff
@@ -238,14 +239,17 @@ func (cp *RestChainProxy) PortalStart(ctx context.Context, privKey *btcec.Privat
 		cp.portalLogs.LogStartTransaction("rest-http")
 
 		msgSeed := cp.portalLogs.GetMessageSeed()
+
 		path := "/" + c.Params("*")
 
 		// TODO: handle contentType, in case its not application/json currently we set it to application/json in the Send() method
 		// contentType := string(c.Context().Request.Header.ContentType())
 		dappID := ExtractDappIDFromFiberContext(c)
+		metricsData := metrics.NewRelayAnalytics(dappID, cp.GetSentry().ChainID, cp.GetSentry().ApiInterface)
```
Comment (ranlavanet) [inline relayer/chainproxy/rest.go:248]: same as grpc.go

File: relayer/chainproxy/portalLogs.go [hunk 20]
```diff
@@ -60,28 +68,35 @@ func (cp *PortalLogs) GetUniqueGuidResponseForError(responseError error, msgSeed
...
+	if pl.newRelicApplication != nil {
+		txn := pl.newRelicApplication.StartTransaction(name)
 		defer txn.End()
 	}
 }
+
+func (pl *PortalLogs) AddMetric(data *metrics.RelayMetrics, isNotSuccessful bool) {
+	if pl.StoreMetricData {
+		data.Success = !isNotSuccessful
+		pl.MetricService.SendData(*data)
```
[the remark this answers is not in the document]
Comment (ranlavanet) [reply relayer/chainproxy/portalLogs.go:100]: edit: call all addmetric using a go routine (from all chainproxy's)

File: relayer/chainproxy/rest.go [hunk 21]
```diff
@@ -271,7 +275,12 @@ func (cp *RestChainProxy) PortalStart(ctx context.Context, privKey *btcec.Privat
 			dappID = strings.ReplaceAll(dappID, "*", "")
 		}
 		utils.LavaFormatInfo("in <<<", &map[string]string{"path": path, "dappID": dappID, "msgSeed": msgSeed})
-		reply, _, err := SendRelay(ctx, cp, privKey, path, query, http.MethodGet, dappID)
+
+		sentry := cp.GetSentry()
+		analytics := metrics.NewRelayAnalytics(dappID, sentry.ChainID, sentry.ApiInterface)
```
Comment (ranlavanet) [inline relayer/chainproxy/rest.go:280]: same as grpc.go :)

File: relayer/chainproxy/tendermintRPC.go [hunk 22]
```diff
@@ -273,12 +274,13 @@ func (cp *tendermintRpcChainProxy) PortalStart(ctx context.Context, privKey *btc
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel() // incase there's a problem make sure to cancel the connection
 			dappID := ExtractDappIDFromWebsocketConnection(c)
-			reply, replyServer, err := SendRelay(ctx, cp, privKey, "", string(msg), http.MethodGet, dappID)
+			metricsData := metrics.NewRelayAnalytics(dappID, cp.GetSentry().ChainID, cp.GetSentry().ApiInterface)
```
Comment (ranlavanet) [inline relayer/chainproxy/tendermintRPC.go:277]: same as grpc.go

File: relayer/chainproxy/tendermintRPC.go [hunk 23]
```diff
@@ -326,7 +328,9 @@ func (cp *tendermintRpcChainProxy) PortalStart(ctx context.Context, privKey *btc
 		msgSeed := cp.portalLogs.GetMessageSeed()
 		dappID := ExtractDappIDFromFiberContext(c)
 		utils.LavaFormatInfo("in <<<", &map[string]string{"seed": msgSeed, "msg": string(c.Body()), "dappID": dappID})
-		reply, _, err := SendRelay(ctx, cp, privKey, "", string(c.Body()), http.MethodGet, dappID)
+		metricsData := metrics.NewRelayAnalytics(dappID, cp.GetSentry().ChainID, cp.GetSentry().ApiInterface)
```
Comment (ranlavanet) [inline relayer/chainproxy/tendermintRPC.go:331]: same as grpc.go

File: relayer/chainproxy/tendermintRPC.go [hunk 24]
```diff
@@ -349,7 +353,9 @@ func (cp *tendermintRpcChainProxy) PortalStart(ctx context.Context, privKey *btc
 		}
 		msgSeed := cp.portalLogs.GetMessageSeed()
 		utils.LavaFormatInfo("urirpc in <<<", &map[string]string{"seed": msgSeed, "msg": path, "dappID": dappID})
-		reply, _, err := SendRelay(ctx, cp, privKey, path+query, "", http.MethodGet, dappID)
+		metricsData := metrics.NewRelayAnalytics(dappID, cp.GetSentry().ChainID, cp.GetSentry().ApiInterface)
```
Comment (ranlavanet) [inline relayer/chainproxy/tendermintRPC.go:356]: same as grpc.go

File: relayer/metrics/metrics_test.go [hunk 25]
```diff
@@ -0,0 +1,237 @@
+package metrics
+
```
Comment (ranlavanet) [inline relayer/metrics/metrics_test.go:2]: add this to e2e.yaml so it will run with the github actions

File: relayer/chainproxy/jsonRPC.go [hunk 26]
```diff
@@ -288,6 +288,10 @@ func (cp *JrpcChainProxy) PortalStart(ctx context.Context, privKey *btcec.Privat
 		}
 		return fiber.ErrUpgradeRequired
 	})
+
+	chainID := cp.GetSentry().ChainID
```
Comment (ranlavanet) [inline relayer/chainproxy/jsonRPC.go:292]: cp.ChainID instead of cp.GetSentry().ChainID

File: relayer/chainproxy/rest.go [hunk 27]
```diff
@@ -234,6 +234,8 @@ func (cp *RestChainProxy) PortalStart(ctx context.Context, privKey *btcec.Privat
 
 	app.Use(favicon.New())
 
+	chainID := cp.GetSentry().ChainID
```
Comment (ranlavanet) [inline relayer/chainproxy/rest.go:237]: cp.ChainID instead of cp.GetSentry().ChainID

File: relayer/chainproxy/tendermintRPC.go [hunk 28]
```diff
@@ -244,6 +244,8 @@ func (cp *tendermintRpcChainProxy) PortalStart(ctx context.Context, privKey *btc
 	//
 	// Setup HTTP Server
 	app := fiber.New(fiber.Config{})
+	chainID := cp.GetSentry().ChainID
```
Comment (ranlavanet) [inline relayer/chainproxy/tendermintRPC.go:247]: cp.ChainID instead of cp.GetSentry().ChainID

**Review by ranlavanet — CHANGES_REQUESTED** [review]

**Review by ranlavanet — APPROVED** [review]

**Pull request merged**